specfile/macro_options.py

```diff
@@ -0,0 +1,523 @@
+
+
+class MacroOptions(collections.abc.MutableMapping):
+    """
+    Class that represents macro options.
+
+    Attributes:
+        optstring: String containing recognized option characters.
+          Option characters are ASCII letters, upper or lower-case.
+          If such a character is followed by a colon, the option
+          requires an argument.
+        defaults: Dict specifying default arguments to options.
+    """
+
+    def __init__(
+        self,
+        tokens: List[Token],
+        optstring: Optional[str] = None,
+        defaults: Optional[Dict[str, Union[bool, int, str]]] = None,
+    ) -> None:
+        """
+        Constructs a `MacroOptions` object.
+
+        Args:
+            tokens: List of tokens in an option string.
+            optstring: String containing recognized option characters.
+              Option characters are ASCII letters, upper or lower-case.
+              If such a character is followed by a colon, the option
+              requires an argument.
+            defaults: Dict specifying default arguments to options.
+
+        Returns:
+            Constructed instance of `MacroOptions` class.
+        """
+        self._tokens = tokens.copy()
+        self.optstring = optstring or ""
+        self.defaults = defaults.copy() if defaults is not None else {}
+
+    def __repr__(self) -> str:
+        tokens = repr(self._tokens)
+        defaults = repr(self.defaults)
+        return f"MacroOptions({tokens}, '{self.optstring}', {defaults})"
+
+    def __str__(self) -> str:
+        return "".join(str(t) for t in self._tokens)
+
+    def _valid_option(self, name: str) -> bool:
+        """
+        Determines if a name represents a recognized option.
+
+        Args:
+            name: Name of the option.
+
+        Returns:
+            True if the option is recognized, otherwise False.
+        """
+        try:
+            # use parent's __getattribute__() so this method can be called from __getattr__()
+            optstring = super().__getattribute__("optstring")
+        except AttributeError:
+            return False
+        return name in optstring.replace(":", "")
+
+    def _requires_argument(self, option: str) -> bool:
+        """
+        Determines if an option requires an argument.
+
+        Args:
+            option: Name of the option.
+
+        Returns:
+            True if the option requires an argument, otherwise False.
+        """
+        i = self.optstring.index(option) + 1
+        return i < len(self.optstring) and self.optstring[i] == ":"
+
+    def _find_option(self, name: str) -> Tuple[int, int]:
+        """
+        Searches for an option in tokens of an option string.
+
+        Args:
+            name: Name of the option.
+
+        Returns:
+            Tuple of indices where the first is the index of a token matching
+            the option and the second is the index of a token matching
+            its argument, or -1 if there is no match.
```

Comment:
Yes, I used a number here so it's easier to use the result in conditions, but thinking about it, with `None` it probably wouldn't be much worse, and it would most likely improve readability.